vslib/src/sai_vs_port.cpp

```diff
@@ -11,7 +12,82 @@ sai_status_t vs_clear_port_all_stats(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
-VS_GENERIC_QUAD(PORT,port);
+sai_status_t vs_create_port(
+            _Out_ sai_object_id_t *port_id,
+            _In_ sai_object_id_t switch_id,
+            _In_ uint32_t attr_count,
+            _In_ const sai_attribute_t *attr_list)
+{
+    MUTEX();
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+
+    /* create port */
+    CHECK_STATUS(meta_sai_create_oid((sai_object_type_t)SAI_OBJECT_TYPE_PORT,
+                port_id,switch_id,attr_count,attr_list,&vs_generic_create));
+
+    attr.id = SAI_PORT_ATTR_ADMIN_STATE;
+    attr.value.booldata = false;     /* default admin state is down as defined in SAI */
+
+    CHECK_STATUS(vs_generic_set(SAI_OBJECT_TYPE_PORT, *port_id, &attr));
+
+    /* create priority groups */
+    const uint32_t port_pgs_count = 8;
+
+    std::vector<sai_object_id_t> pgs;
+
+    for (uint32_t i = 0; i < port_pgs_count; ++i)
+    {
+        sai_object_id_t pg_id;
+
+        CHECK_STATUS(vs_generic_create(SAI_OBJECT_TYPE_INGRESS_PRIORITY_GROUP, &pg_id, switch_id, 0, NULL));
+
+        pgs.push_back(pg_id);
+    }
+
+    attr.id = SAI_PORT_ATTR_NUMBER_OF_INGRESS_PRIORITY_GROUPS;
+    attr.value.u32 = port_pgs_count;
+
+    CHECK_STATUS(vs_generic_set(SAI_OBJECT_TYPE_PORT, *port_id, &attr));
+
+    attr.id = SAI_PORT_ATTR_INGRESS_PRIORITY_GROUP_LIST;
+    attr.value.objlist.count = port_pgs_count;
+    attr.value.objlist.list = pgs.data();
+
+    CHECK_STATUS(vs_generic_set(SAI_OBJECT_TYPE_PORT, *port_id, &attr));
+
+    /* create qos queues */
+    const uint32_t port_qos_queues_count = 20;
+
+    std::vector<sai_object_id_t> queues;
+
+    for (uint32_t i = 0; i < port_qos_queues_count; ++i)
+    {
+        sai_object_id_t queue_id;
+
+        CHECK_STATUS(vs_generic_create(SAI_OBJECT_TYPE_QUEUE, &queue_id, switch_id, 0, NULL));
+
+        queues.push_back(queue_id);
+    }
+
+    attr.id = SAI_PORT_ATTR_QOS_NUMBER_OF_QUEUES;
+    attr.value.u32 = port_qos_queues_count;
+
+    CHECK_STATUS(vs_generic_set(SAI_OBJECT_TYPE_PORT, *port_id, &attr));
+
+    attr.id = SAI_PORT_ATTR_QOS_QUEUE_LIST;
+    attr.value.objlist.count = port_qos_queues_count;
+    attr.value.objlist.list = queues.data();
+
+    CHECK_STATUS(vs_generic_set(SAI_OBJECT_TYPE_PORT, *port_id, &attr));
+
+    return SAI_STATUS_SUCCESS;
+}
+
+VS_REMOVE(PORT,port);
```

Comment:
if you do all these things in the create_port, then you need also do reverse things in the remove_port.